### PR TITLE
[FIXED JENKINS-42795] Incorrect type mapping of href deserialization in team objects

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-42795](https://issues.jenkins-ci.org/browse/JENKINS-42795)

@reviewbybees

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudTeam.java [hunk 1]
```diff
@@ -65,12 +66,12 @@ public void setDisplayName(String displayName) {
...
     @Override
     @JsonIgnore
-    public Map<String, BitbucketHref> getLinks() {
+    public Map<String, List<BitbucketHref>> getLinks() {
         return links;
     }
 
     @JsonIgnore
-    public void setLinks(Map<String, BitbucketHref> links) {
+    public void setLinks(Map<String, List<BitbucketHref>> links) {
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudTeam.java]: Is it an override as well? If no, needs to retain the compatibility

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/repository/BitbucketServerProject.java [hunk 2]
```diff
@@ -66,24 +67,13 @@ public void setDisplayName(String displayName) {
...
-            result.put(entry.getKey(), entry.getValue()[0]);
-        }
-        return result;
+    @JsonIgnore
+    public Map<String, List<BitbucketHref>> getLinks() {
+        return links;
     }
 
-    // Do not call this setLinks or Jackson will have issues
-    public void links(Map<String, BitbucketHref> links) {
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/repository/BitbucketServerProject.java]: same

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java [hunk 3]
```diff
@@ -327,11 +323,15 @@ private String bitbucketUrl() {
...
+    private static String getLink(Map<String, List<BitbucketHref>> links, String name) {
         if (links == null) {
             return null;
         }
-        BitbucketHref href = links.get(name);
+        List<BitbucketHref> hrefs = links.get(name);
+        if (hrefs == null || hrefs.isEmpty()) {
+            return null;
+        }
+        BitbucketHref href = hrefs.get(0);
```
Comment (amuniz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java]: Is this going to contain always one single element?

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudTeam.java (hunk 1 above)
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/repository/BitbucketCloudTeam.java]: I have changed the method signature in the interface. Not used outside of this plugin so safe to do

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java (hunk 3 above)
Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java]: It's going to have at least one element if not empty (see 4 lines above)

**Comment by stephenc** [comment]
@oleg-nenashev said:

>  I am not sure if there are legitimate use-cases outside the plugin,

We are not encouraging people to use the current API. It is really an internal only separation of two internal only implementations.

At some point in the future we will probably redo the API into a better clearer separation of concerns

**Comment by oleg-nenashev** [comment]
@stephenc Maybe you want to mark all the API classes as Restricted then

**Comment by stephenc** [comment]
@oleg-nenashev happy now?

**Comment by oleg-nenashev** [comment]
@stephenc yep, removing the bug.
Does BCT support package annotations BTW?

**Comment by stephenc** [comment]
@oleg-nenashev 
> Does BCT support package annotations BTW?

Not my problem

**Pull request merged**